bridge-worker/server.go

```diff
@@ -0,0 +1,195 @@
+	chainID *big.Int
+}
+
+func NewServer(
+	logger log.Logger,
+	port uint16,
+	chainID *big.Int,
+) *Server {
+	s := &Server{
+		logger:  logger,
+		chainID: chainID,
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("POST /submit-signature", s.submitSignature)
+
+	s.server = &http.Server{
+		Addr:              fmt.Sprintf(":%d", port),
+		Handler:           mux,
+		ReadHeaderTimeout: time.Second,
+	}
+
+	return s
+}
+
+func (s *Server) Start() {
+	s.logger.Info("http server starting", "address", s.server.Addr)
+	go func() {
+		if err := s.server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			// should we panic here? likely yes
+			s.logger.Error("server failed to start", "error", err)
+		}
+	}()
+}
+
+func (s *Server) Stop(ctx context.Context) error {
+	return s.server.Shutdown(ctx)
+}
+
+func (s *Server) submitSignature(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	req, err := readSubmitSignatureRequest(r)
+	if err != nil {
+		writeError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	if err := req.Validate(); err != nil {
+		writeError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	address, err := s.recoverAddress(req.Entry, req.Signature)
+	if err != nil {
+		writeError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	// now do call some other service which will
+	// process the asset unlock if the address is a
+	// valid signer
+	_ = address
+
+	writeSuccess(w, http.StatusAccepted)
+}
+
+func (s *Server) recoverAddress(entry *types.AssetsUnlocked, signature string) (common.Address, error) {
+	abiEncoded, err := abiEncodeAttestationWithChainID(entry, s.chainID)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	// Apply the same TextHash transformation
+	hash := accounts.TextHash(abiEncoded)
+
+	// Decode the signature from hex
+	signatureBytes, err := hexutil.Decode(signature)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	// Recover the public key from the signature
+	publicKeyBytes, err := crypto.SigToPub(hash, signatureBytes)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	// Convert public key to address
+	return crypto.PubkeyToAddress(*publicKeyBytes), nil
+}
+
+func readSubmitSignatureRequest(r *http.Request) (*types.SubmitSignatureRequest, error) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		return nil, errors.New("failed to read request body")
+	}
+	defer r.Body.Close()
+
+	req := types.SubmitSignatureRequest{}
+	if err := json.Unmarshal(body, &req); err != nil {
+		return nil, errors.New("invalid json format")
+	}
+
+	return &req, nil
+}
+
+func writeError(w http.ResponseWriter, err error, status int) {
+	w.WriteHeader(status)
+	_ = json.NewEncoder(w).Encode(types.SubmitSignatureResponse{
+		Error:   err.Error(),
+		Success: false,
+	})
+}
+
+func writeSuccess(w http.ResponseWriter, status int) {
+	w.WriteHeader(status)
+	_ = json.NewEncoder(w).Encode(types.SubmitSignatureResponse{
+		Success: true,
+	})
+}
```

Comment:
Those are specific to our endpoint, let's either make them private functions inside the handler or extend their names to contain the endpoint name. We will likely have other endpoints in the future here, hence my point.

Reply:
69c5cc3d22996cd9ecce0663bbf8899e924f4491